contracts/modules/cover/Cover.sol

```diff
@@ -225,109 +244,124 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard {
-
-      _coverSegments[coverId].push(
-        CoverSegment(
-          coverAmountInCoverAsset.toUint96(), // amount
-          uint32(block.timestamp + 1), // start
-          SafeUintCast.toUint32(params.period), // period
-          _productTypes[_products[params.productId].productType].gracePeriodInDays,
-          globalRewardsRatio
-        )
-      );
+    // TODO: implement using buckets
+    totalActiveCoverInAsset[params.coverAsset] += coverAmountInCoverAsset;
 
+    if (amountDueInNXM > 0) {
       retrievePayment(
-        premiumInNXM,
+        amountDueInNXM,
         params.paymentAsset,
         params.maxPremiumInAsset,
         params.commissionRatio,
         params.commissionDestination
       );
-
-      // TODO: implement using buckets
-      totalActiveCoverInAsset[params.coverAsset] += coverAmountInCoverAsset;
-
-      emit CoverEdited(coverId, params.productId, segmentId, msg.sender, params.ipfsData);
     }
+
+    emit CoverEdited(coverId, params.productId, segmentId, msg.sender, params.ipfsData);
   }
 
+  function requestAllocation(
+    AllocationRequest memory allocationRequest,
+    PoolAllocationRequest[] memory poolAllocationRequests,
+    uint nxmPriceInCoverAsset,
+    uint segmentId
+  ) internal returns (
+    uint totalCoverAmountInCoverAsset,
+    uint totalAmountDueInNXM
+  ) {
 
-  function isAllowedPool(uint productId, uint poolId) external view returns (bool) {
+    RequestAllocationVariables memory vars = RequestAllocationVariables(0, 0, 0, 0);
+    uint totalCoverAmountInNXM;
 
-    if (!_products[productId].useFixedPrice) {
-      // no pool id restrictions
-      return true;
-    }
-    bool poolIsAllowed;
-    uint[] memory allowedPoolsForProduct = allowedPools[productId];
-    for (uint i = 0; i < allowedPoolsForProduct.length; i++) {
-      if (allowedPoolsForProduct[i] == poolId) {
-        poolIsAllowed = true;
-        break;
+    vars.previousPoolAllocationsLength = segmentId > 0
+      ? coverSegmentAllocations[allocationRequest.coverId][segmentId - 1].length
+      : 0;
+
+    for (uint i = 0; i < poolAllocationRequests.length; i++) {
+
+      // TODO: add a flag in PoolAllocationRequest to skip certain pools to avoid repricing
+      // TODO: poolAllocationRequests might have repeated pools, is this gameable?
+
+      // if there is a previous segment and this index is present on it
+      if (vars.previousPoolAllocationsLength > i) {
+
+        PoolAllocation memory previousPoolAllocation =
+          coverSegmentAllocations[allocationRequest.coverId][segmentId - 1][i];
+
+        // poolAllocationRequests must match the pools in the previous segment
+        require(previousPoolAllocation.poolId == poolAllocationRequests[i].poolId, "Cover: Unexpected pool id");
+
+        vars.previousPremiumInNXM = previousPoolAllocation.premiumInNXM;
+        vars.refund =
+          previousPoolAllocation.premiumInNXM
```

Comment:
noting that this is a change in how refund is computed.

We used to express refund in cover Asset (Eg. DAI, ETH etc.)

now it's expressed in NXM

Reply:
Good point. Will raise with Hugh but it won't be fixed in this PR since it's a spec issue not a code issue.